lib/interactor.rb

```diff
@@ -112,8 +112,16 @@ def initialize(context = {})
   #
   # Returns nothing.
   def run
-    run!
-  rescue Failure
+    catch(:early_return) do
+      with_hooks do
+        call
+        context.called!(self)
+      end
+    end
+    context.rollback! if context.failure?
+  rescue
+    context.rollback!
+    raise
```

Comment:
In general, I need to stare at and think about this for a little while. I'm not as familiar with `throw` and `catch` as I am with `raise` and `rescue`, so we just need to think through the edge cases.